idseq_dag/steps/blast_contigs.py

```diff
@@ -433,6 +439,47 @@ def get_top_m8_nr(orig_m8, blast_top_m8):
             if top_line is not None:
                 top_m8f.write(top_line)  # Unify reranked formats for NT and NR
 
+
+    @staticmethod
+    def filter_and_group_hits_by_query(blast_output, min_alignment_length, min_pident):
+        # Filter and group results by query, yielding one result group at a time.
+        # A result group consists of all hits for a query, grouped by subject.
+        # Please see comment in get_top_m8_nt for more context.
+        current_query = None
+        current_query_hits = defaultdict(list)
+        previously_seen_queries = set()
+        # Please see comments explaining the definition of "hsp" elsewhere in this file.
+        for hsp in m8.parse_tsv(blast_output, m8.BLAST_OUTPUT_NT_SCHEMA):
+            # filter local alignment HSPs based on minimum length and sequence similarity
+            if hsp["length"] < min_alignment_length:
+                continue
+            if hsp["pident"] < min_pident:
+                continue
+            query, subject = hsp["qseqid"], hsp["sseqid"]
+            if query != current_query and current_query_hits:
+                assert query not in previously_seen_queries, "blastn output appears out of order, please resort by (qseqid, sseqid, score)"
+                previously_seen_queries.add(query)
+                yield current_query_hits
+                current_query = query
+                current_query_hits = defaultdict(list)
+            current_query_hits[subject].append(hsp)
+        if current_query_hits:
+            yield current_query_hits
+
+
+    @staticmethod
+    def optimal_hit_for_each_query(blast_output, min_alignment_length, min_pident):
+        # Yield the optimal hit for each query, from amongst the many subjects hit by the query.  The hit is a list of fragments that together maximize a certain score.  Please see comment in get_top_m8_nt for more context.
+        for query_hits in PipelineStepBlastContigs.filter_and_group_hits_by_query(blast_output, min_alignment_length, min_pident):
+            best_hit = None
+            for _subject, hit_fragments in query_hits.items():
+                hit = BlastCandidate(hit_fragments)
+                hit.optimize()
+                if not best_hit or best_hit.total_score < hit.total_score:
+                    best_hit = hit
+            yield best_hit.summary()
```

Comment:
This optimization logic is also exactly the same, and just yields the optimal hits one at a time instead of storing them in a data structure.    In this way they are streamed into the output file below as soon as they are generated, and the memory needed to hold them is then released.